x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java

```diff
@@ -46,14 +47,29 @@
 
     public static final IndicesPermission NONE = new IndicesPermission();
 
-    private final ConcurrentMap<String, Predicate<String>> allowedIndicesMatchersForAction = new ConcurrentHashMap<>();
+    private final ConcurrentMap<String, Predicate<IndexAbstraction>> allowedIndicesMatchersForAction = new ConcurrentHashMap<>();
 
     private final Group[] groups;
 
     public IndicesPermission(Group... groups) {
         this.groups = groups;
     }
 
+    public static Predicate<String> indexMatcher(Collection<String> ordinaryIndices, Collection<String> restrictedIndices) {
+        final Predicate<String> namePredicate;
+        if (restrictedIndices.isEmpty()) {
+            namePredicate = indexMatcher(ordinaryIndices)
+                    .and(index -> false == RestrictedIndicesNames.isRestricted(index));
+        } else if (ordinaryIndices.isEmpty()) {
+            namePredicate = indexMatcher(restrictedIndices);
+        } else {
+            namePredicate = indexMatcher(restrictedIndices)
+                    .or(indexMatcher(ordinaryIndices)
+                            .and(index -> false == RestrictedIndicesNames.isRestricted(index)));
```

Comment:
It's a minor thing, but I have a mild dislike of that repeated `isRestricted` logic.

We could do something like:
```
    final Predicate<String> namePredicate;
    if (ordinaryIndices.isEmpty()) {
        namePredicate = indexMatcher(restrictedIndices);
    } else {
        namePredicate = indexMatcher(ordinaryIndices)
                .and(index -> false == RestrictedIndicesNames.isRestricted(index));
        if (restrictedIndices.isEmpty() == false) {
            namePredicate = indexMatcher(restrictedIndices).or(namePredicate);
        }
    }
```